### PR TITLE
SubCommand_Find: Add finding enchantments on books and gear

### PR DESCRIPTION
**Pull request by SarahGreyWolf**
This feature allows users to use `/qs find` to search for enchantments on both Enchanted Books and Gear such as Swords, Pickaxes and Armour.

I previously made this for [PotatoCraft-Studio/QuickShop-Reremake](https://github.com/PotatoCraft-Studio/QuickShop-Reremake) and have seen it get a lot of use on minecraft servers.

An example of its usage would be a shop selling a Diamond Sword with Sharpness 5, performing `/qs find sharpness` would direct you towards that shop, it is a very useful tool for servers that have large market areas, where just searching for Enchanted Book is not good enough and the option that allows just the first Enchantment is equally not good enough for multi-enchant books, as well as when you wish to find upgraded gear being sold at said market places.

### CODE REVIEW

File: quickshop-bukkit/src/main/java/com/ghostchu/quickshop/util/Util.java [hunk 1]
```diff
@@ -497,6 +497,86 @@ public static boolean isEmptyComponent(@Nullable Component component) {
...
+    /**
+     * Find a string in a component
+     *
+     * @param component The component to check
+     * @param find The string to check for
+     * @return A boolean of whether the component contains the string
+     */
+    @NotNull
+    public static boolean findStringInComponent(@NotNull Component component, @NotNull String find) {
+        String plainText = PlainTextComponentSerializer.plainText().serialize(component).toLowerCase();
```
Comment (creatorfromhell) [inline quickshop-bukkit/src/main/java/com/ghostchu/quickshop/util/Util.java]: plainText String can be made final.

File: quickshop-bukkit/src/main/java/com/ghostchu/quickshop/util/Util.java [hunk 2]
```diff
@@ -497,6 +497,86 @@ public static boolean isEmptyComponent(@Nullable Component component) {
...
+
+        ItemMeta meta = itemStack.getItemMeta();
+        if (meta instanceof EnchantmentStorageMeta enchantmentStorageMeta && enchantmentStorageMeta.hasStoredEnchants()) {
+            for (Map.Entry<Enchantment, Integer> entry : enchantmentStorageMeta.getStoredEnchants().entrySet()) {
+                Component name = enchantmentDataToComponent(entry.getKey(), entry.getValue());
+                enchants.add(name);
+            }
+        } else {
+            for (Map.Entry<Enchantment, Integer> entry : meta.getEnchants().entrySet()) {
+                Component name = enchantmentDataToComponent(entry.getKey(), entry.getValue());
```
Comment (creatorfromhell) [inline quickshop-bukkit/src/main/java/com/ghostchu/quickshop/util/Util.java]: Same thing here, a few things can be noted as final just for readability.
- enchants list
- component name
- ItemMeta

File: quickshop-bukkit/src/main/java/com/ghostchu/quickshop/util/Util.java [hunk 3]
```diff
@@ -507,18 +587,8 @@ public static Component getFirstEnchantmentName(@NotNull EnchantmentStorageMeta
...
-        } catch (Throwable throwable) {
-            name = MsgUtil.setHandleFailedHover(null, Component.text(entry.getKey().getKey().getKey()));
-            plugin.logger().warn("Failed to handle translation for Enchantment {}", entry.getKey().getKey(), throwable);
-        }
-        if (entry.getValue() == 1 && entry.getKey().getMaxLevel() == 1) {
-            return name;
-        } else {
-            return name.append(Component.text(" " + RomanNumber.toRoman(entry.getValue())));
-        }
+        Component name = enchantmentDataToComponent(entry.getKey(), entry.getValue());
```
Comment (creatorfromhell) [inline quickshop-bukkit/src/main/java/com/ghostchu/quickshop/util/Util.java]: Would it break code here to just return the enchantmentDataToComponent versus storing it as a variable?

**Review by creatorfromhell — CHANGES_REQUESTED** [review]
Overall looks good, just some minor nitpicks that will be part of the future code refactor style guidelines.

**Comment by SarahGreyWolf** [comment]
I tend to forget about final, bad habit, would be helpful to have a style guide

**Comment by creatorfromhell** [comment]
> I tend to forget about final, bad habit, would be helpful to have a style guide

indeed, I have to get everything in the core styled according to the style guide I have in Intellij, then I'll have to add the github bot for it.

**Review by creatorfromhell — APPROVED** [review]

**Pull request merged**